File: bindings/python/examples/reconstruct.py
"""Reconstruct a 3D ultrasound volume from simulated plane-wave data.

Given channel data, element positions, a reconstruction grid, and
transmit parameters, ffdas.das produces a volumetric image. The
simulation section below generates synthetic data using a trefoil
knot phantom; skip it if you have your own channel data.
"""

import cupy as cp
import matplotlib.pyplot as plt

import ffdas


cp.random.seed(0)

sound_speed = 1540.0
center_freq = 3.08e6
sampling_freq = center_freq  # complex baseband
n_samples = 512

# probe: 32x32 matrix array in the xy-plane at z=0
pitch = 500e-6
el = (cp.arange(32, dtype=cp.float32) - 15.5) * pitch
ex, ey = cp.meshgrid(el, el, indexing="ij")  # type: ignore
channel_pos = cp.column_stack([ex.ravel(), ey.ravel(), cp.zeros_like(ex).ravel()])
n_channels = channel_pos.shape[0]

xmin, xmax = -0.008, 0.008
ymin, ymax = -0.008, 0.008
zmin, zmax = 0.010, 0.026


# simulate rf data for a normal-incidence plane wave


def trefoil(t):
    return cp.stack(
        [
            (cp.sin(t) + 2 * cp.sin(2 * t)) * 0.1 * (xmax - xmin),
            -cp.sin(3 * t) * 0.15 * (ymax - ymin),
            (cp.cos(t) - 2 * cp.cos(2 * t)) * 0.1 * (zmax - zmin) + 0.5 * (zmin + zmax),
        ],
        axis=-1,
    )


batch_size = 128
n_scatterers = 4096

tube_sigma = 0.0001
bg_sigma = 0.008

p = 0.3
t = 2 * cp.pi * cp.random.rand(batch_size, n_scatterers, dtype=cp.float32)
displacement = cp.random.normal(
    scale=cp.random.choice([tube_sigma, bg_sigma], size=(1, n_scatterers, 1), p=[p, 1-p]), 
    size=(batch_size, n_scatterers, 3),
    dtype=cp.float32,  # type: ignore
)
scatter_pos = trefoil(t) + displacement

freqs = cp.fft.fftfreq(n_samples, d=1.0 / sampling_freq).astype(cp.float32) + center_freq
wavenums = (-2 * cp.pi * freqs / sound_speed).astype(cp.float32)
sigma_f = 0.6 * center_freq / (2 * cp.sqrt(2 * cp.log(2)))
pulse = cp.exp(-0.5 * ((freqs - center_freq) / sigma_f) ** 2).astype(cp.complex64)

# plane wave: all channels transmit simultaneously (zero delay)
channel_delay = cp.zeros(channel_pos.shape[0], dtype=cp.float32)
transmission = pulse * cp.exp(-2j * cp.pi * freqs * channel_delay[:, None])

tx = ffdas.greens(channel_pos, wavenums, transmission, scatter_pos)
rx = ffdas.greens(scatter_pos, wavenums, tx, channel_pos)
rf = cp.fft.ifft(rx, axis=-1).astype(cp.complex64).conj()

# rf: (batch, channels, samples). add the sequence dimension (one
# transmission here; see compound.py for multi-angle compounding)
rf = rf[:, :, None, :]  # (batch, channels, 1, samples)


# reconstruction grid: 64^3 voxels centered on the phantom
nz, ny, nx = 64, 64, 64
x = cp.linspace(xmin, xmax, nx, dtype=cp.float32)
y = cp.linspace(ymin, ymax, ny, dtype=cp.float32)
z = cp.linspace(zmin, zmax, nz, dtype=cp.float32)
zz, yy, xx = cp.meshgrid(z, y, x, indexing="ij")  # type: ignore
voxel_pos = cp.stack([xx, yy, zz], axis=-1)

# all positions passed to das are in sampling wavelengths (c/fs)
ks = sampling_freq / sound_speed

# transmit offset: one-way plane-wave delay from z=0 to each voxel.
# shape is (n_transmissions, nz, ny, nx), matching the sequence dimension
offsets = voxel_pos[None, ..., 2] * ks

# transmit apodization (uniform for a plane wave)
weights = cp.ones_like(offsets)

with ffdas.utils.Timer() as t:
    image = ffdas.das(
        rf,
        channel_pos * ks,
        voxel_pos * ks,
        offsets,
        weights,
        srcdir=None,
        wavenum=-2 * cp.pi * center_freq / sampling_freq,
    )
print(f"das (srcdir=None): {nz}x{ny}x{nx}, {channel_pos.shape[0]} ch, batch {batch_size}: {t.elapsed_ms():.1f} ms")

# element directivity: (channels, 4) where columns 0-2 are the element
# normal and column 3 is the cosine cutoff. voxels outside the cone
# receive zero weight from that channel
srcdir = cp.zeros((channel_pos.shape[0], 4), dtype=cp.float32)
srcdir[:, 2] = 1.0  # normals point in +z
srcdir[:, 3] = 0.707  # 60 degree half-angle

with ffdas.utils.Timer() as t:
    image = ffdas.das(
        rf,
        channel_pos * ks,
        voxel_pos * ks,
        offsets,
        weights,
        srcdir=srcdir,
        wavenum=-2 * cp.pi * center_freq / sampling_freq,
    )
print(f"das (with srcdir): {nz}x{ny}x{nx}, {channel_pos.shape[0]} ch, batch {batch_size}: {t.elapsed_ms():.1f} ms")


magnitude = cp.abs(image[0])

mip_xz = magnitude.max(axis=1)
db_xz = 20 * cp.log10(mip_xz / mip_xz.max() + 1e-10)

mip_yz = magnitude.max(axis=2)
db_yz = 20 * cp.log10(mip_yz / mip_yz.max() + 1e-10)

fig, axes = plt.subplots(1, 2, figsize=(10, 8), sharey=True)

axes[0].imshow(
    cp.asnumpy(db_xz),
    extent=[float(x[0]) * 1e3, float(x[-1]) * 1e3, float(z[-1]) * 1e3, float(z[0]) * 1e3],
    cmap="gray",
    vmin=-24,
    vmax=0,
    aspect="equal",
)
axes[0].set_xlabel("x [mm]")
axes[0].set_ylabel("z [mm]")
axes[0].set_title("xz max projection")

axes[1].imshow(
    cp.asnumpy(db_yz),
    extent=[float(y[0]) * 1e3, float(y[-1]) * 1e3, float(z[-1]) * 1e3, float(z[0]) * 1e3],
    cmap="gray",
    vmin=-24,
    vmax=0,
    aspect="equal",
)
axes[1].set_xlabel("y [mm]")
axes[1].set_title("yz max projection")

plt.tight_layout()
plt.savefig("reconstruct.png", dpi=150)

File: bindings/matlab/examples/reconstruct.m
% Reconstruct a 3D ultrasound volume from simulated plane-wave data.
%
% Given channel data, element positions, a reconstruction grid, and
% transmit parameters, ffdas.das produces a volumetric image. The
% simulation section below generates synthetic data using a trefoil
% knot phantom; skip it if you have your own channel data.

rng(0);

sound_speed = 1540.0;
center_freq = 3.08e6;
sampling_freq = center_freq;
n_samples = 512;

pitch = 500e-6;
el = gpuArray(single((0:31) - 15.5) * pitch);
[ex, ey] = ndgrid(el, el);
n_channels = 1024;
channel_pos = gpuArray(single([ex(:)'; ey(:)'; zeros(1, n_channels)]));

xmin = -0.008; xmax = 0.008;
ymin = -0.008; ymax = 0.008;
zmin = 0.010;  zmax = 0.026;


% simulate rf data for a normal-incidence plane wave.
% see simulation.py for a walkthrough of greens propagation

batch_size = 128;
n_scatterers = 16384;
n_knot = 8192;

t = 2 * pi * rand(1, n_knot, "single", "gpuArray");
knot_pos = trefoil(t) + randn(3, n_knot, "single", "gpuArray") * 0.00015;
background_pos = [ ...
    xmin + (xmax - xmin) * rand(1, n_scatterers - n_knot, "single", "gpuArray"); ...
    ymin + (ymax - ymin) * rand(1, n_scatterers - n_knot, "single", "gpuArray"); ...
    zmin + (zmax - zmin) * rand(1, n_scatterers - n_knot, "single", "gpuArray")];

scatter_pos = cat(2, knot_pos, background_pos);
scatter_values = cat(2, ...
    rand(1, n_knot, batch_size, "single", "gpuArray"), ...
    rand(1, n_scatterers - n_knot, batch_size, "single", "gpuArray"));

k_fft = gpuArray(single([0:n_samples/2-1, -n_samples/2:-1]'));
freqs = k_fft * sampling_freq / n_samples + center_freq;
wavenums = single(-2 * pi) * freqs / sound_speed;
sigma_f = 0.6 * center_freq / (2 * sqrt(2 * log(2)));
pulse = complex(exp(-0.5 * ((freqs - center_freq) / sigma_f).^2));

% plane wave: all channels transmit simultaneously (zero delay)
channel_delay = zeros(1, n_channels, "single", "gpuArray");
transmission = pulse .* exp(-2j * pi * freqs .* channel_delay);

tx = ffdas.greens(channel_pos, wavenums, transmission, scatter_pos);
rx = ffdas.greens(scatter_pos, wavenums, tx .* scatter_values, channel_pos);
rf = conj(ifft(rx, [], 1));

% (samples, channels, batch) -> (samples, 1, channels, batch)
rf = reshape(rf, n_samples, 1, [], batch_size);


% reconstruction grid: 64^3 voxels centered on the phantom
nz = 64; ny = 64; nx = 64;
x = gpuArray(single(linspace(xmin, xmax, nx)));
y = gpuArray(single(linspace(ymin, ymax, ny)));
z = gpuArray(single(linspace(zmin, zmax, nz)));
[xx, yy, zz] = ndgrid(x, y, z);
voxel_pos = zeros(3, nx, ny, nz, "single", "gpuArray");
voxel_pos(1,:,:,:) = xx;
voxel_pos(2,:,:,:) = yy;
voxel_pos(3,:,:,:) = zz;

ks = sampling_freq / sound_speed;

% transmit offset: one-way plane-wave delay from z=0 to each voxel
% (nx, ny, nz, 1) — trailing dimension is the sequence axis
offsets = squeeze(voxel_pos(3,:,:,:)) * ks;
offsets = reshape(offsets, nx, ny, nz, 1);
weights = ones(size(offsets), "single", "gpuArray");

wavenum = single(-2 * pi * center_freq / sampling_freq);

timer = ffdas.utils.Timer();
timer.start();
image = ffdas.das( ...
    rf, channel_pos * ks, voxel_pos * ks, offsets, weights, [], wavenum);
timer.stop();
fprintf("das (srcdir=[]): %dx%dx%d, %d ch, batch %d: %.1f ms\n", ...
    nz, ny, nx, n_channels, batch_size, timer.elapsed_ms());

% element directivity: (4, channels), rows 1-3 are the normal, row 4
% is the cosine of the sensitivity half-angle
srcdir = zeros(4, n_channels, "single", "gpuArray");
srcdir(3,:) = 1.0;
srcdir(4,:) = 0.707;  % ~45 degree half-angle

timer2 = ffdas.utils.Timer();
timer2.start();
image = ffdas.das( ...
    rf, channel_pos * ks, voxel_pos * ks, offsets, weights, srcdir, wavenum);
timer2.stop();
fprintf("das (with srcdir): %dx%dx%d, %d ch, batch %d: %.1f ms\n", ...
    nz, ny, nx, n_channels, batch_size, timer2.elapsed_ms());


magnitude = abs(image(:,:,:,1));

mip_xz = squeeze(max(magnitude, [], 2));
db_xz = 20 * log10(mip_xz / max(mip_xz(:)) + 1e-10);

mip_yz = squeeze(max(magnitude, [], 1));
db_yz = 20 * log10(mip_yz / max(mip_yz(:)) + 1e-10);

figure;
tiledlayout(1, 2);

nexttile;
imagesc(gather(x) * 1e3, gather(z) * 1e3, gather(db_xz)');
colormap("gray"); clim([-24 0]);
xlabel("x [mm]"); ylabel("z [mm]");
title("xz max projection");
axis image;

nexttile;
imagesc(gather(y) * 1e3, gather(z) * 1e3, gather(db_yz)');
colormap("gray"); clim([-24 0]);
xlabel("y [mm]");
title("yz max projection");
axis image;

exportgraphics(gcf, "reconstruct.png", Resolution=150);


function pos = trefoil(t)
    pos = cat(1, ...
        (sin(t) + 2 * sin(2*t)) * 0.0016, ...
        -sin(3*t) * 0.0024, ...
        (cos(t) - 2 * cos(2*t)) * 0.0016 + 0.018);
end
